### PR TITLE
[ads] Add SmartNTT virtual prefs and condition matchers

### PR DESCRIPTION
**Pull request by tmancey**
Resolves https://github.com/brave/brave-browser/issues/41927

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

Adds support for the following virtual pref condition matchers:
```
{
   "[virtual]:browser": {
      "build_channel": "nightly",
      "version": "131.1.73.0"
   },
   "[virtual]:operating_system": {
      "locale": {
         "language": "en",
         "region": "KY"
      },
      "name": "Mac OS X"
   },
   "[virtual]:search_engine": {
      "default_name": "Brave"
   },
   "[virtual]:skus": {
      "production": {
         "talk.brave.com": {
            "created_at": "2023-05-05T12:26:57",
            "expires_at": "2024-05-04T13:44:55",
            "last_paid_at": "2024-04-05T13:44:55",
            "status": "paid"
         },
         "vpn.brave.com": {
            "created_at": "2023-02-23T21:49:34",
            "expires_at": "2024-09-24T21:49:57",
            "last_paid_at": "2024-08-25T21:49:57",
            "status": "canceled"
         }
      }
   }
}
```

where `operating_system.name` can be "Windows", "Mac OS X", "Linux", "Android", "iOS", or "Unknown".

where SKUs status can be either `beta`, `trial`, `paid`, or `canceled`.

Example to match if running on a desktop operating system:
```
"conditionMatchers": [
  {
    "condition": "^(Windows|Mac OS X|Linux)$",
    "prefPath": "[virtual]:operating_system|name"
  }
]
```
Example to match if a VPN premium subscription is canceled and expired 30 or more days ago:
```
"conditionMatchers": [
  {
    "condition": "[T]≥30",
    "prefPath": "[virtual]:skus|production|vpn.brave.com|expires_at"
  },
  {
    "condition": "canceled",
    "prefPath": "[virtual]:skus|production|vpn.brave.com|status"
  }
]
```
--

`[numerical operator]:number` condition matcher which supports the following operators:
- 'R=': Equal
- 'R≠': Not equal
- 'R>': Greater than
- 'R≥': Greater than or equal to
- 'R<': Less than
- 'R≤': Less than or equal to

This matcher triggers an ad based on when a real number (integers or fractions) stored at "prefPath". For instance, the example below will serve an ad only if the value stored at "foo.bar" is not equal to 3:
```
"conditionMatchers": [
  {
    "condition": "[R≠]:3",
    "prefPath": "foo.bar"
  }
]
```

Also, changes epoch condition matchers from `[?]:*` to `[T?]:*`. Epoch condition matchers now support IOS 8601 date and time stored as string values at the pref path, i.e. `2024-09-30T19:00:00T`, `30 Sept 2024 19:00:00 EST` etc.

### CODE REVIEW

File: browser/brave_ads/ads_service_delegate.h [hunk 1]
```diff
@@ -76,6 +81,7 @@ class AdsServiceDelegate : public AdsService::Delegate {
  private:
   raw_ptr<Profile> profile_ = nullptr;
   raw_ptr<PrefService> local_state_ = nullptr;
+  std::string application_locale_;
```
Comment (aseren) [inline browser/brave_ads/ads_service_delegate.h:84]: nit: const std::string application_locale_;

Comment (tmancey) [reply browser/brave_ads/ads_service_delegate.h]: Fixed. Thanks

**Review by aseren — APPROVED** [review]

File: browser/brave_ads/ads_service_delegate.cc [hunk 2]
```diff
@@ -53,6 +59,77 @@ AdsServiceDelegate::AdsServiceDelegate(
...
+  return base::UTF16ToUTF8(default_search_engine_name);
+}
+
+base::Value::Dict AdsServiceDelegate::GetSkus() const {
+  base::Value::Dict dict;
+
+  const base::Value::Dict& skus_state_dict =
+      local_state_->GetDict(skus::prefs::kSkusState);
+
+  for (const auto sku_state_dict : skus_state_dict) {
```
[the remark this answers is not in the document]
Comment (tmancey) [reply browser/brave_ads/ads_service_delegate.cc:76]: Fixed but cannot be a ref

**Comment by btlechowski** [comment]
Verified with

Brave | 1.73.62 Chromium: 131.0.6778.14 (Official Build) nightly (64-bit)
-- | --
Revision | dcd982b6f2279349c9b97ffe99963388a479769f
OS | Linux

Verified test plan from https://github.com/brave/brave-core/pull/26284

#### [virtual]:browser|build_channel

<details>

```
                    "conditionMatchers": [{
                            "prefPath": "[virtual]:browser|build_channel",
                            "condition": "nightly"
                        }
                    ],
```

Verified ad is shown

</details>

#### [virtual]:operating_system

<details>

```
                    "conditionMatchers": [{
                            "prefPath": "[virtual]:operating_system|name",
                            "condition": "Linux"
                        }
                    ],
```

Verified ad is shown

</details>


#### [virtual]:search_engine

<details>

```
                    "conditionMatchers": [{
                            "prefPath": "[virtual]:search_engine|default_name",
                            "condition": "Brave"
                        }
                    ],
```

Verified ad is shown

</details>

#### virtual]:skus

<details>

```
                    "conditionMatchers": [{
                            "condition": "[T=]:0",
                            "prefPath": "[virtual]:skus|production|talk.brave.com|created_at"
                        }
                    ],
```

Verified ad is shown

</details>

#### [numerical operator]:number

<details>

```
                    "conditionMatchers": [{
                            "prefPath": "tab_stats.window_count_max",
                            "condition": "[R>]:1"
                        }
                    ],
```

Verified ad is shown

</details>
